### PR TITLE
HDF5 file for binding

### PR DESCRIPTION
**Pull request by simonvh**
The output of `ananse binding` is now an HDF5 file, that contains the binding predictions as well as the factor activity, with the following layout:

```
/FACTOR1/
/FACTOR2/
/FACTORN/
/_index/
/_factor_activity/
```

The advantage is three-fold:
1) Save diskspace (aggressive compression is used).
2) Data can be efficiently read per chromosome. This should solve the high memory use of `ananse network`, which turns out to be very difficult to solve otherwise.
3) The info from `factor_activity.tsv` is now also stored in the h5, which means it moves around with the file, and there's no implicit dependency on this file being present.

### CODE REVIEW

**Comment by siebrenf** [comment]
If we're gonna store the binding file in a non-text format, we can also use a more compact data structure!

right now `binding.tsv` looks like this:
```
TF1    region_1    1
TF1    region_2    2
TF2    region_1    5
TF2    region_2    6
```

why not make this a table?
```
          TF1  TF2
region_1   1    5
region_2   2    6
```

**Comment by siebrenf** [comment]
unless that's exactly what you do with `hdf.put(key="_index", value=proba.index.to_series(), format="table")`

File: ananse/network.py [hunk 1]
```diff
@@ -368,54 +373,84 @@ def aggregate_binding(
...
-        if tfs is not None:
-            ddf = ddf[ddf["factor"].isin(tfs)]
-
-        # Merge binding information with gene information.
-        # This may be faster than creating index on enhancer first, but need to check!
-        tmp = ddf.merge(gene_df, left_on="enhancer", right_index=True)
-
-        # Remove everything with weight 0
-        tmp = tmp[tmp["weight"] > 0]
+        # TODO: This is hacky (depending on "_"), however the hdf.keys() method is
```
Comment (siebrenf) [inline ananse/network.py:378]: if "\_" feels hacky, you could prefix the files with "ANANSE_". Either way, speed is king, so I'd agree.

File: ananse/network.py [hunk 2]
```diff
@@ -648,4 +682,7 @@ def __del__(self):
 
         for fname in self._tmp_files:
             if os.path.exists(fname):
-                os.unlink(fname)
+                if os.path.isdir(fname):
```
Comment (siebrenf) [inline ananse/network.py:685]: `shutil.rmtree(fname, ignore_errors=True)` should cover files and dirs

File: ananse/network.py [hunk 3]
```diff
@@ -368,54 +373,84 @@ def aggregate_binding(
...
+            )
 
-        logger.info("Grouping by tf and target gene...")
-        # Using one column that combines TF and target as dask cannot handle MultiIndex
-        tmp["tf_target"] = tmp["factor"] + "_" + tmp["gene"]
+            gene_df = gene_df.dropna()
+
+            bp = pd.DataFrame(index=enhancers[idx].index)
+
+            for tf in tqdm(tfs):
```
Comment (siebrenf) [inline ananse/network.py:422]: `tqdm(tfs, total=len(tfs), desc="Loading", unit_scale=1, unit=" TFs")` etc

**Review by siebrenf — APPROVED** [review]

File: ananse/network.py [hunk 4]
```diff
@@ -368,54 +373,84 @@ def aggregate_binding(
...
+            bp = bp.mul(gene_df["weight"], axis=0)
+            bp["gene"] = gene_df["gene"]
+            tmp = bp.groupby("gene")
+            if combine_function == "mean":
+                tmp = tmp.mean()
+            elif combine_function == "max":
+                tmp = tmp.max()
+            elif combine_function == "sum":
+                tmp = tmp.sum()
+            tmp = tmp.reset_index().melt(
```
Comment (Maarten-vd-Sande) [inline ananse/network.py:441]:
```
else:
    raise NotImplementedError
```

**Review by Maarten-vd-Sande — APPROVED** [review]
Looks good! 

Haven't looked at most of what it actually is doing, this stuff is always hard to follow (more comments always help :innocent: ), just if the code seems okay :)

On a side-note, I had quite some troubles with hdf5 and qnorm, apparantly it only accepts a max of 512 columns. My _impression_ so far (disclaimer: minimal experience) is that hdf5 is getting "old", and probably using feather in the future is easier for maintainability reasons.

**Comment by simonvh** [comment]
> unless that's exactly what you do with `hdf.put(key="_index", value=proba.index.to_series(), format="table")`

Yes! The enhancer locations are only stored once. At the moment, TFs are stored separately, which allows to also retrieve them separately. All to save memory later down the line.

**Comment by simonvh** [comment]
> On a side-note, I had quite some troubles with hdf5 and qnorm, apparantly it only accepts a max of 512 columns. My _impression_ so far (disclaimer: minimal experience) is that hdf5 is getting "old", and probably using feather in the future is easier for maintainability reasons.

The limit seems to be ~2000-3000 columns (depending on type and column name?). Should be enough for TFs. In any case, I think [zarr](https://zarr.readthedocs.io/en/stable/) would be a good way forward. Feather is nice, but takes up a lot of space (and can't store multiple tables in one file).

File: ananse/network.py (hunk 4 above)
Comment (simonvh) [reply ananse/network.py:441]: It's already checked earlier in the function.
